packages/shylu/shylu_dd/frosch/src/SchwarzOperators/FROSch_CoarseOperator_def.hpp

```diff
@@ -67,6 +67,191 @@ namespace FROSch {
         CoarseSolver_.reset();
     }
 
+    //##############CoarseSolve Repeated Map##############################
+    //this is only called if an additional level is applied and ZoltanDual is chosen
+    //for distribution
+    template<class SC,class LO, class GO, class NO>
+    int CoarseOperator<SC,LO,GO,NO>::buildGlobalGraph(Teuchos::RCP<DDInterface<SC,LO,GO,NO> > theDDInterface_)
+    {
+      FROSCH_TIMER_START_LEVELID(buildGlobalGraphTime,"CoarseOperator::buildGlobalGraph");
+      std::map<GO,int> rep;
+      Teuchos::Array<GO> entries;
+      IntVec2D conn;
+      InterfaceEntityPtrVec ConnVec;
+      int connrank;
+      //get connected subdomains
+      {
+         theDDInterface_->identifyConnectivityEntities();
+         EntitySetConstPtr Connect=  theDDInterface_->getConnectivityEntities();
+         Connect->buildEntityMap(theDDInterface_->getNodesMap());
+         ConnVec = Connect->getEntityVector();
+         connrank = Connect->getEntityMap()->getComm()->getRank();
+       }
+       GO ConnVecSize = ConnVec.size();
+       conn.resize(ConnVecSize);
+       {
+         if (ConnVecSize>0) {
+           for(GO i = 0;i<ConnVecSize;i++) {
+               conn[i] = ConnVec[i]->getSubdomainsVector();
+               for (int j = 0; j<conn[i].size(); j++) rep.insert(std::pair<GO,int>(conn.at(i).at(j),connrank));
+           }
+           for (auto& x: rep) {
+               entries.push_back(x.first);
+           }
+         }
+       }
+
+       Teuchos::RCP<Xpetra::Map<LO,GO,NO> > GraphMap = Xpetra::MapFactory<LO,GO,NO>::Build(this->K_->getMap()->lib(),-1,1,0,this->K_->getMap()->getComm());
+
+       //UN maxNumElements = -1;
+       maxNumNeigh_ = -1;
+       UN numElementsLocal = entries.size();
+       reduceAll(*this->MpiComm_,Teuchos::REDUCE_MAX,numElementsLocal,Teuchos::ptr(&maxNumNeigh_));
+       SubdomainConnectGraph_ = Xpetra::CrsGraphFactory<LO,GO,NO>::Build(GraphMap,maxNumNeigh_);
+       SubdomainConnectGraph_->insertGlobalIndices(GraphMap->getComm()->getRank(),entries());
+
+       return 0;
+    }
+
+    template <class SC,class LO, class GO,class NO>
+    int CoarseOperator<SC,LO,GO,NO>::buildCoarseGraph()
+    {
+      //bring graph to correct communictaor
+      FROSCH_TIMER_START_LEVELID(buildCoarseGraphTime,"CoarseOperator::buildCoarseGraph");
+
+		  GraphPtr TestGraph2 =  Xpetra::CrsGraphFactory<LO,GO,NO>::Build(MLGatheringMaps_[1],maxNumNeigh_);;
```

Comment:
Can you add some comment if you introduce generic variables like "`TestGraph2`"?